runtime/freelist.c

```diff
@@ -597,27 +1755,147 @@ void caml_make_free_blocks (value *p, mlsize_t size, int do_merge, int color)
     }else{
       sz = size;
     }
-    *(header_t *)p =
-      Make_header (Wosize_whsize (sz), 0, color);
-    if (do_merge) caml_fl_merge_block (Val_hp (p));
+    if (do_merge){
+      mlsize_t wosz = Wosize_whsize (sz);
+      if (wosz <= BF_NUM_SMALL){
+        color = Caml_white;
+      }else{
+        color = Caml_blue;
+      }
+      *(header_t *)p = Make_header (wosz, 0, color);
+      bf_insert_remnant (Val_hp (p));
+    }else{
+      *(header_t *)p = Make_header (Wosize_whsize (sz), 0, color);
+    }
     size -= sz;
     p += sz;
   }
 }
 
-void caml_set_allocation_policy (uintnat p)
+/*********************** policy selection *****************************/
+
+typedef enum caml_policy_t {
+  policy_next_fit = 0,
+  policy_first_fit = 1,
+  policy_best_fit = 2,
+} caml_policy_t;
+
+caml_policy_t caml_allocation_policy = policy_next_fit;
```

Comment:
This change from `uintnat` to `caml_policy_t` needs to be reflected in `gc_ctrl.c` -- at the moment `caml_gc_get` reads the allocation policy and 7 other bytes of memory to return to the user.

Reply:
Color me surprised.  I assume this is the relevant line of `caml_gc_get`?
```
  Store_field (res, 6, Val_long (caml_allocation_policy));              /* a */
```
The friendly C compiler will read `sizeof(caml_policy_t)` bytes of memory, then convert the resulting value to an OCaml integer via `Val_long`.

Reply:
I think it is the:
```
extern uintnat caml_allocation_policy
```
line further up in `gc_ctrl.c` that is causing problems.
I defer to you on the vagaries of C compilers and I haven't checked that changing this line fixes it, but the behaviour is definitely consistent with reading the wrong size from memory.

Reply:
Ouch!  This is awful indeed.  Thanks for pointing out.

Reply:
Sigh. That's what I get for trying to clean up the code... Fixed.